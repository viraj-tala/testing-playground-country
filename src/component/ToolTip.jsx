import React from 'react';

export const Tooltip = ({ text, position, children }) => {
    const getPositionStyles = () => {
        switch (position) {
            case 'top':
                return 'bottom-full left-1/2 transform -translate-x-1/2';
            case 'right':
                return 'top-1/2 left-full transform -translate-y-1/2';
            case 'bottom':
                return 'top-full left-1/2 transform -translate-x-1/2';
            case 'left':
                return 'top-1/2 right-full transform -translate-y-1/2';
            default:
                return 'bottom-full left-1/2 transform -translate-x-1/2';
        }
    };

    return (
        <div className="relative inline-block">
            <div className="group relative">
                {children}
                <div
                    className={`opacity-0 invisible group-hover:opacity-100 group-hover:visible absolute z-10 bg-slate-800 text-white py-2 px-4 rounded-md text-center whitespace-nowrap ${getPositionStyles()}`}
                >
                    {text}
                </div>
            </div>
        </div>
    );
};