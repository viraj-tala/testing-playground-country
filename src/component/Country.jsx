import { useEffect, useState } from "react";
import { Link, useParams } from "react-router-dom";
import Loader from "../loader/Loader";
import { MoreOutlined, HomeOutlined } from "@material-ui/icons";
import { Tooltip } from "./ToolTip";

const Country = () => {
    let { countryCode } = useParams();
    const [country, setCountry] = useState(null);
    const [borders, setBorders] = useState(null);
    useEffect(() => {
        const fetchCountryDetails = async () => {
            try {
                const res = await fetch(`https://restcountries.com/v3.1/alpha/${countryCode}`);
                const countryData = await res.json();
                setCountry(countryData[0]);

                // Fetch neighboring countries
                const borderCodes = countryData[0].borders || [];
                const borderCountries = await Promise.all(
                    borderCodes.map(async (code) => {
                        const borderRes = await fetch(`https://restcountries.com/v3.1/alpha/${code}`);
                        return borderRes.json();
                    })
                );
                setBorders(borderCountries);
            } catch (error) {
                console.log("Error fetching country details:", error);
            }
        };

        fetchCountryDetails();
    }, [countryCode]);

    if (!country) {
        return <Loader />;
    }

    return (
        <div className="m-10">
            <div className="flex m-14 mb-0 justify-end hover:text-blue-500">
                <Tooltip text='Back To Home' position='bottom'>
                    <Link to={`/`}>
                        <HomeOutlined />
                    </Link>
                </Tooltip>
            </div>
            <h4 className="mb-10 flex justify-center text-5xl font-extrabold leading-none tracking-tight text-gray-900 ">{country?.name?.common}</h4>

            <div className="flex items-center justify-center">
                <div className="w-full flex flex-col items-center justify-center max-w-sm p-6 bg-white border border-gray-200 rounded-lg shadow hover:bg-gray-100 dark:bg-gray-800 dark:border-gray-700 dark:hover:bg-gray-700">
                    <img
                        className="w-28 h-20 flex rounded-xl"
                        src={country.flags.png}
                        alt={country.flags.alt}
                    />
                    <div className="m-3">
                        <h4 className="text-2xl font-bold tracking-tight text-gray-200">
                            {country?.name?.common}
                        </h4>
                        <p className="font-normal flex justify-center text-gray-700 dark:text-gray-300">
                            {country.region}
                        </p>
                    </div>
                    <div className="w-full flex justify-between">
                        <div className="">
                            <p className="font-normal text-gray-300">
                                {country.population}
                            </p>
                            <div className="overview_label text-14 text-gray-400 text-secondary">
                                Population
                            </div>
                        </div>
                        <div className="">
                            <p className="font-normal text-gray-300">
                                {country.area}
                            </p>
                            <div className="overview_label text-14 text-gray-400 text-secondary">
                                Area
                            </div>
                        </div>
                    </div>
                </div>
            </div>
            <div className="">
                <div className="details_panel bg-light box-shadow rounded-8 p-6 ">
                    <h4 className="text-xl p-2 font-semibold text-gray-200 mb-10 pb-2 border-b-2 rounded-xl dark:bg-gray-800 dark:border-gray-700 dark:hover:bg-gray-700">
                        Details
                    </h4>
                    <ul className="space-y-4">
                        <li className="flex items-start pb-2 border-b-2 justify-between">
                            <span className="font-medium text-gray-900">
                                Capital
                            </span>
                            <span className="text-gray-700 font-semibold ">
                                {country.capital[0]}
                            </span>
                        </li>
                        {country.languages && <li className="flex items-start justify-between pb-2 border-b-2">
                            <span className="font-medium text-gray-900">
                                Languages
                            </span>
                            <div className="flex flex-wrap">
                                {Object.entries(country.languages).map(([code, name]) => (
                                    <div key={code} className="mr-2 mb-2 font-semibold text-gray-800">
                                        <span className="font-semibold text-gray-700">
                                            {code}:
                                        </span>
                                        {name}
                                    </div>
                                ))}
                            </div>
                        </li>}
                        {country.currencies && <li className="flex items-start justify-between pb-2 border-b-2">
                            <span className="font-medium text-gray-900">
                                Currencies
                            </span>
                            <div className="flex flex-wrap">
                                {Object.entries(country.currencies).map(([code, currency]) => (
                                    <div key={code} className="mr-2 mb-2">
                                        <span className="font-semibold text-gray-700">
                                            {currency.name} ({currency.symbol})
                                        </span>
                                    </div>
                                ))}
                            </div>
                        </li>}
                        {country.name.nativeName && <li className="flex items-start justify-between pb-2 border-b-2">
                            <span className="font-medium text-gray-900">
                                Native name
                            </span>
                            <span className="text-gray-700 font-semibold ">
                                {Object.entries(country.name.nativeName).map(([key, value]) => (
                                    <div key={key} className="mr-2 mb-2">
                                        <span className="font-semibold text-gray-700">
                                            {key}: {value.common}
                                        </span>
                                    </div>
                                ))}
                            </span>
                        </li>}
                        {country.gini && <li className="flex items-start justify-between pb-2 border-b-2">
                            <span className="font-medium text-gray-900">
                                Gini
                            </span>
                            <span className="text-gray-700 font-semibold">
                                {Object.entries(country.gini).map(([key, value]) => (
                                    <div key={key} className="mr-2 mb-2">
                                        <span className="font-semibold text-gray-700">
                                            {key && key}: {value}%
                                        </span>
                                    </div>
                                ))}
                            </span>
                        </li>}
                        {borders?.length && <li className="flex items-start justify-between pb-2 border-b-2">
                            <span className="font-medium flex h-full text-gray-900">
                                Neighboring Countries
                            </span>
                            <div className="flex flex-wrap flex-col gap-2">
                                {borders.map((border) =>
                                    border[0] ? (
                                        <div
                                            key={border[0].name?.common}
                                            className="mr-2 mb-2 flex gap-2"

                                        >
                                            <img
                                                className="w-12 h-8 rounded-md"
                                                src={border[0].flags?.png}
                                                alt={border[0].flags?.alt}
                                            />
                                            <div className="flex justify-end  w-full gap-2  hover:text-blue-500 ">
                                                <span className="font-semibold text-gray-700  ">
                                                    {border[0].name?.common}
                                                </span>
                                                <Tooltip text='More Info' position='bottom'>
                                                    {border[0].cioc && <Link to={`/country/${border[0].cioc}`}>
                                                        <MoreOutlined />
                                                    </Link>}
                                                </Tooltip>
                                            </div>
                                        </div>
                                    ) : <></>
                                )}
                            </div>
                        </li>}
                    </ul>
                </div>
            </div>

        </div >
    );
};

export default Country;
