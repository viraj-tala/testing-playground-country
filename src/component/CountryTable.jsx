import React, { useState } from "react";
import { KeyboardArrowDownRounded, KeyboardArrowUpRounded, MoreOutlined } from "@material-ui/icons";
import { Link } from "react-router-dom";
import Loader from "../loader/Loader";
import { Tooltip } from "./ToolTip";


const orderBy = (countries, value, direction) => {
    if (direction === "ascending") {
        return [...countries].sort((a, b) => (a[value] > b[value] ? 1 : -1));
    } else if (direction === "descending") {
        return [...countries].sort((a, b) => (a[value] > b[value] ? -1 : 1));
    }

    return countries;
};

const SortArrow = ({ direction }) => {
    if (!direction) {
        return <></>;
    }

    if (direction === "descending") {
        return (
            <div className="heading_arrow">
                <KeyboardArrowDownRounded color="inherit" />
            </div>
        );
    } else if (direction === "ascending") {
        return (
            <div className="heading_arrow">
                <KeyboardArrowUpRounded color="inherit" />
            </div>
        );
    }
};

const CountriesTable = ({ countries }) => {
    const [direction, setDirection] = useState("ascending");
    const [value, setValue] = useState("");

    const orderedCountries = orderBy(countries, value, direction);

    const switchDirection = () => {
        if (!direction) {
            setDirection("descending");
        } else if (direction === "descending") {
            setDirection("ascending");
        } else {
            setDirection(null);
        }
    };

    const setValueAndDirection = (value) => {
        switchDirection();
        setValue(value);
    };

    const countriesList = (
        <tbody className="overflow-auto">
            {orderedCountries.map(({ name, flags, population, area, region, subregion, cioc }) => (
                <tr key={name.common} role="row" className="row pb-10">

                    <td className="pl-3 ">
                        <img src={flags.svg} alt={name.common} className="h-20 w-30 rounded" />
                    </td>
                    <td className="pl-3">
                        {name.common}
                    </td>
                    <td className="pl-7">{population}</td>
                    <td className="pl-6">{area || 0}</td>
                    <td className="pl-7">{region || 0}</td>
                    <td className="pl-3">{subregion || 0}</td>
                    <td className="pl-8 hover:text-blue-500">
                        <Tooltip text='More Info' position='bottom'>
                            {cioc && <Link to={`/country/${cioc}`}>
                                <MoreOutlined />
                            </Link>}
                        </Tooltip>
                    </td>

                </tr>
            ))}
        </tbody>
    );

    return (
        <div>
            {!orderedCountries.length ? <Loader />
                :
                <div className="relative m-10 overflow-x-auto shadow-md sm:rounded-lg">
                    <table role="table" className="w-full text-sm text-left rtl:text-right text-gray-500 dark:text-gray-400">
                        <thead className="text-xs text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-400">
                            <tr>
                                <th scope="col" className="px-6 py-3">
                                    Flag
                                </th>
                                <th scope="col" className="px-6 py-3">
                                    <button className="flex items-center" onClick={() => setValueAndDirection("name")}>
                                        <div>Name</div>
                                        {value === "name" && <SortArrow direction={direction} />}
                                    </button>
                                </th>
                                <th scope="col" className="px-6 py-3">
                                    <button className="flex items-center"
                                        onClick={() => setValueAndDirection("population")}
                                    >
                                        <div>Population</div>
                                        {value === "population" && <SortArrow direction={direction} />}
                                    </button>
                                </th>
                                <th scope="col" className="px-6 py-3">
                                    <button className="flex items-center"
                                        onClick={() => setValueAndDirection("area")}
                                    >
                                        <div>
                                            Area (km<sup style={{ fontSize: "0.75rem" }}>2</sup>)
                                        </div>
                                        {value === "area" && <SortArrow direction={direction} />}
                                    </button>
                                </th>
                                <th scope="col" className="px-6 py-3">
                                    <button className="flex items-center"
                                        onClick={() => setValueAndDirection("region")}
                                    >
                                        <div>Region</div>
                                        {value === "region" && <SortArrow direction={direction} />}
                                    </button>
                                </th>

                                <th scope="col" className="px-6 py-3">
                                    <button className="flex items-center"
                                        onClick={() => setValueAndDirection("subregion")}
                                    >
                                        <div>Subregion</div>
                                        {value === "subregion" && <SortArrow direction={direction} />}
                                    </button>
                                </th>

                                <th scope="col" className="px-6 py-3">
                                    <div>Details</div>
                                </th>
                            </tr>
                        </thead>
                        {countriesList}
                    </table>
                </div>
            }
        </div>
    );
};

export default CountriesTable