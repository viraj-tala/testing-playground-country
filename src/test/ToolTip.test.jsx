/* eslint-disable testing-library/render-result-naming-convention */
import React from 'react';
import { render, screen } from './test.utils';
import userEvent from '@testing-library/user-event';
import { Tooltip } from '../component/ToolTip';

describe('Tooltip', () => {

    it('should match the snapshot', () => {
        const element = render(<Tooltip position='top' text='Top Tooltip' >
            <span>Hover over me</span>
        </Tooltip>)
        expect(element).toMatchSnapshot();
    })

    it('should call onChange when the input value changes', async () => {
        render(<Tooltip position='top' text='Top Tooltip' >
            <span>Hover over me</span>
        </Tooltip>);

        const triggerElement = screen.getByText('Hover over me');
        const tooltipElement = screen.getByText('Top Tooltip');

        expect(triggerElement).toBeInTheDocument();
        expect(tooltipElement).toBeInTheDocument();

        userEvent.hover(triggerElement);

        expect(tooltipElement).toHaveClass('group-hover:visible');
    });
});
