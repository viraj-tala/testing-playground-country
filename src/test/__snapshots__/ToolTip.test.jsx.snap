// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative inline-block"
      >
        <div
          class="group relative"
        >
          <span>
            Hover over me
          </span>
          <div
            class="opacity-0 invisible group-hover:opacity-100 group-hover:visible absolute z-10 bg-slate-800 text-white py-2 px-4 rounded-md text-center whitespace-nowrap bottom-full left-1/2 transform -translate-x-1/2"
          >
            Top Tooltip
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative inline-block"
    >
      <div
        class="group relative"
      >
        <span>
          Hover over me
        </span>
        <div
          class="opacity-0 invisible group-hover:opacity-100 group-hover:visible absolute z-10 bg-slate-800 text-white py-2 px-4 rounded-md text-center whitespace-nowrap bottom-full left-1/2 transform -translate-x-1/2"
        >
          Top Tooltip
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
